sandbox/apparmor/notify/listener/listener_test.go

```diff
@@ -0,0 +1,728 @@
+	os.Setenv("PROMPT_NOTIFY_PATH", fakePath)
+
+	l, err = listener.Register()
+	c.Assert(err, IsNil)
+
+	c.Assert(outputOverridePath, Equals, fakePath)
+
+	err = l.Close()
+	c.Assert(err, IsNil)
+}
+
+func (*listenerSuite) TestRegisterErrors(c *C) {
+	restoreOpen := listener.MockOsOpen(func(name string) (*os.File, error) {
+		return nil, os.ErrNotExist
+	})
+	defer restoreOpen()
+
+	l, err := listener.Register()
+	c.Assert(l, IsNil)
+	c.Assert(err, Equals, listener.ErrNotSupported)
+
+	customError := errors.New("custom error")
+
+	restoreOpen = listener.MockOsOpen(func(name string) (*os.File, error) {
+		return nil, customError
+	})
+	defer restoreOpen()
+
+	l, err = listener.Register()
+	c.Assert(l, IsNil)
+	c.Assert(err, ErrorMatches, fmt.Sprintf("cannot open %q: %v", notify.SysPath, customError))
+
+	restoreOpen = listener.MockOsOpen(func(name string) (*os.File, error) {
+		placeholderSocket, err := unix.Socket(unix.AF_UNIX, unix.SOCK_STREAM, 0)
+		c.Assert(err, IsNil)
+		placeholderFile := os.NewFile(uintptr(placeholderSocket), name)
+		return placeholderFile, nil
+	})
+	defer restoreOpen()
+
+	restoreIoctl := listener.MockNotifyIoctl(func(fd uintptr, req notify.IoctlRequest, buf notify.IoctlRequestBuffer) ([]byte, error) {
+		return nil, customError
+	})
+	defer restoreIoctl()
+
+	l, err = listener.Register()
+	c.Assert(l, IsNil)
+	c.Assert(err, ErrorMatches, fmt.Sprintf("cannot notify ioctl to modeset user on %q: %v", notify.SysPath, customError))
+
+	restoreOpen = listener.MockOsOpen(func(name string) (*os.File, error) {
+		badFd := ^uintptr(0)
+		fakeFile := os.NewFile(badFd, name)
+		return fakeFile, nil
+	})
+	defer restoreOpen()
+
+	restoreIoctl = listener.MockNotifyIoctl(func(fd uintptr, req notify.IoctlRequest, buf notify.IoctlRequestBuffer) ([]byte, error) {
+		return nil, nil
+	})
+	defer restoreIoctl()
+
+	l, err = listener.Register()
+	c.Assert(l, IsNil)
+	c.Assert(err, ErrorMatches, fmt.Sprintf("cannot register epoll on %q: bad file descriptor", notify.SysPath))
+}
+
+// An expedient abstraction over notify.MsgNotificationFile to allow defining
+// MsgNotificationFile structs using value literals.
+type msgNotificationFile struct {
+	// MsgHeader
+	Length  uint16
+	Version uint16
+	// MsgNotification
+	NotificationType notify.NotificationType
+	Signalled        uint8
+	NoCache          uint8
+	ID               uint64
+	Error            int32
+	// msgNotificationOpKernel
+	Allow uint32
+	Deny  uint32
+	Pid   uint32
+	Label uint32
+	Class uint32
+	Op    uint32
+	// msgNotificationFileKernel
+	SUID uint32
+	OUID uint32
+	Name uint32
+}
+
+func (*listenerSuite) TestRun(c *C) {
+	sockFdChan, restoreOpen := listener.MockOsOpenWithSockets()
+	defer restoreOpen()
+
+	restoreIoctl := listener.MockNotifyIoctlWithReadWrite()
+	defer restoreIoctl()
+
+	var t tomb.Tomb
+	l, err := listener.Register()
+	c.Assert(err, IsNil)
+	defer func() {
+		c.Assert(l.Close(), IsNil)
+		c.Assert(t.Wait(), Equals, listener.ErrClosed)
```

Comment:
Thanks for catching this before. Tracking down and fixing this flakiness led to other flakiness, but I believe I've finally resolved it all now.

Reply:
I seem to have spoken to soon, regarding the most recent test failures.

Assuming the problem is similar to those I've been working on all day, the easy fix would be to use the new channel-based mocking instead of socket writing. But I'll confirm tomorrow that that is indeed the case, and fix it either way.